pkg/openservicebroker/client/client.go

```diff
@@ -0,0 +1,318 @@
+	req, err := http.NewRequest(http.MethodGet, c.root+"/v2/service_instances/"+instanceID+"/last_operation?operation="+string(operation), nil)
+	if err != nil {
+		return nil, err
+	}
+	req = req.WithContext(ctx)
+
+	req.Header.Add(openservicebroker.XBrokerAPIVersion, openservicebroker.APIVersion)
+
+	resp, err := c.cli.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.Header.Get("Content-Type") != "application/json" {
+		return nil, errors.New("invalid response")
+	}
+
+	d := json.NewDecoder(resp.Body)
+	if resp.StatusCode == http.StatusOK || resp.StatusCode == http.StatusGone {
+		var r *openservicebroker.LastOperationResponse
+		err = d.Decode(&r)
+		if err != nil {
+			return nil, err
+		}
+		return r, nil
+	}
+
+	var r *openservicebroker.ErrorResponse
+	err = d.Decode(&r)
+	if err != nil {
+		return nil, err
+	}
+	return nil, errors.New(r.Description)
+}
+
+func (c *client) WaitForOperation(ctx context.Context, instanceID string, operation openservicebroker.Operation) (openservicebroker.LastOperationState, error) {
+	done := ctx.Done()
+	for {
+		op, err := c.LastOperation(ctx, instanceID, operation)
+		if err != nil {
+			return openservicebroker.LastOperationStateFailed, err
+		}
+
+		if op.State != openservicebroker.LastOperationStateInProgress {
+			return op.State, nil
+		}
+
+		select {
+		case <-done:
+			return openservicebroker.LastOperationStateFailed, ctx.Err()
+		default:
+		}
+
+		time.Sleep(1 * time.Second)
+	}
+}
+
+func (c *client) Bind(ctx context.Context, instanceID, bindingID string, breq *openservicebroker.BindRequest) (*openservicebroker.BindResponse, error) {
+	pr, pw := io.Pipe()
+	go func() {
+		e := json.NewEncoder(pw)
+		pw.CloseWithError(e.Encode(breq))
+	}()
+
+	req, err := http.NewRequest(http.MethodPut, c.root+"/v2/service_instances/"+instanceID+"/service_bindings/"+bindingID, pr)
+	if err != nil {
+		return nil, err
+	}
+	req = req.WithContext(ctx)
+
+	req.Header.Add(openservicebroker.XBrokerAPIVersion, openservicebroker.APIVersion)
+	req.Header.Add("Content-Type", "application/json")
+
+	resp, err := c.cli.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.Header.Get("Content-Type") != "application/json" {
+		return nil, errors.New("invalid response")
+	}
+
+	d := json.NewDecoder(resp.Body)
+	if resp.StatusCode == http.StatusOK || resp.StatusCode == http.StatusCreated {
+		var r *openservicebroker.BindResponse
+		err = d.Decode(&r)
+		if err != nil {
+			return nil, err
+		}
+		return r, nil
+	}
+
+	var r *openservicebroker.ErrorResponse
+	err = d.Decode(&r)
+	if err != nil {
+		return nil, err
+	}
+	return nil, errors.New(r.Description)
+}
+
+func (c *client) Unbind(ctx context.Context, instanceID, bindingID string) error {
+	req, err := http.NewRequest(http.MethodDelete, c.root+"/v2/service_instances/"+instanceID+"/service_bindings/"+bindingID, nil)
```

Comment:
Not an excuse, but note that we vendor a lot of client libraries that follow this pattern, including both docker client libraries.

I'm not clear what fix will satisfy you here, whether starting from Golang primitives or using some existing library, unless it's something like ensuring instanceID and bindingID match `^[-A-Za-z0-9]+$`